File: app/src/main/java/org/schabi/newpipe/streams/io/SharpInputStream.java
package org.schabi.newpipe.streams.io;

import androidx.annotation.NonNull;

import java.io.IOException;
import java.io.InputStream;

/**
 * Simply wraps a readable {@link SharpStream} allowing it to be used with built-in Java stuff that
 * supports {@link InputStream}.
 */
public class SharpInputStream extends InputStream {
    private final SharpStream stream;

    public SharpInputStream(final SharpStream stream) throws IOException {
        if (!stream.canRead()) {
            throw new IOException("SharpStream is not readable");
        }
        this.stream = stream;
    }

    @Override
    public int read() throws IOException {
        return stream.read();
    }

    @Override
    public int read(@NonNull final byte[] b) throws IOException {
        return stream.read(b);
    }

    @Override
    public int read(@NonNull final byte[] b, final int off, final int len) throws IOException {
        return stream.read(b, off, len);
    }

    @Override
    public long skip(final long n) throws IOException {
        return stream.skip(n);
    }

    @Override
    public int available() {
        final long res = stream.available();
        return res > Integer.MAX_VALUE ? Integer.MAX_VALUE : (int) res;
    }

    @Override
    public void close() {
        stream.close();
    }
}

File: app/src/main/java/org/schabi/newpipe/streams/io/SharpInputStream.kt
package org.schabi.newpipe.streams.io

import java.io.InputStream
import okio.IOException

/**
 * Simply wraps a readable [SharpStream] allowing it to be used with built-in Java stuff that
 * supports [InputStream].
 */
class SharpInputStream
@Throws(IOException::class)
constructor(
    private val stream: SharpStream
) : InputStream() {

    init {
        if (!stream.canRead()) {
            throw IOException("SharpStream is not readable")
        }
    }

    @Throws(IOException::class)
    override fun read(): Int {
        return stream.read()
    }

    @Throws(IOException::class)
    override fun read(b: ByteArray): Int {
        return stream.read(b)
    }

    @Throws(IOException::class)
    override fun read(b: ByteArray, off: Int, len: Int): Int {
        return stream.read(b, off, len)
    }

    @Throws(IOException::class)
    override fun skip(n: Long): Long {
        return stream.skip(n)
    }

    override fun available(): Int {
        val res = stream.available()
        return if (res > Int.MAX_VALUE) Int.MAX_VALUE else res.toInt()
    }

    override fun close() {
        stream.close()
    }
}
